Implement more conversions from JsonDynamicValue

gvkries opened the pull request:
Reintroduces conversions to fundamental runtime types that were previously supported by `Newtonsoft.Json` but became unavailable following the upgrade to `System.Text.Json`. By incorporating these conversions, we restore the ability to handle dynamic types effectively, maintaining a seamless transition and compatibility across different JSON processing scenarios.

Fixes #15762

MikeAlhayek reviewed (approved):
This LGTM. 

@sebastienros can you please have a look at this one?

gvkries commented:
I fixed an regression of this PR, because overloads of `ToString` were missing. 
Also I added the `Value` property back to `JsonDynamicValue` in case some one has used something like `contentItem.Content.MyPart.myField.Text.Value` instead of just `contentItem.Content.MyPart.myField.Text`. That was possible with Newtonsoft Json as well.

sebastienros commented on test/OrchardCore.Tests/Data/ContentItemTests.cs:
```diff
@@ -155,7 +155,70 @@ public void ShouldDeserializeListContentPart()
+            contentItem.Alter<MyPart>(x =>
+            {
+                x.GetOrCreate<MyDateTimeField>("myField");
+                x.Alter<MyDateTimeField>("myField", f => f.Value = new DateTime(2024, 1, 1, 10, 42, 0));
```
I'd like to see the same test with a UTC datetime and how it gets back when deserialized.

gvkries replied:
Not really related to this PR, but I added two tests.

sebastienros commented on src/OrchardCore/OrchardCore.Abstractions/Json/Dynamic/JsonDynamicValue.cs:
```diff
@@ -1,35 +1,282 @@
+        {
+        }
+
+        public override DynamicMetaObject BindConvert(ConvertBinder binder)
```
What C# is invoking this code?

gvkries replied:
I attempted to come up with a good comment 😇

gvkries replied:
I don't think that makes sense, it's an overridden runtime method which is documented on the base class.

sebastienros commented on src/OrchardCore/OrchardCore.Abstractions/Json/Dynamic/JsonDynamicValue.cs:
```diff
@@ -1,35 +1,282 @@
+        {
+            var targetType = binder.Type;
+
+            var castMethod = typeof(JsonDynamicValue).GetMethods(BindingFlags.Public | BindingFlags.Static)
```
Does it mean that each "cast" will do some reflection and then Expression interpretation? These expressions should be cached.

gvkries replied:
Yes, that is true for dynamic invocations anyway. I added a cache for the reflection information. But the dynamic performance is bad anyway, so I originally thought that would not be necessary.

MikeAlhayek commented:
@gvkries can you please address @sebastienros and I'll merge this PR.

MikeAlhayek commented:
@sebastienros you good with the latest changes in this PR? Can we merge it?

sebastienros commented on src/OrchardCore/OrchardCore.Abstractions/Json/Dynamic/JsonDynamicValue.cs:
```diff
@@ -1,35 +1,297 @@
+
+    private sealed class JsonDynamicMetaObject : DynamicMetaObject
+    {
+        private static readonly Dictionary<Type, MethodInfo> _cachedReflectionInfo = typeof(JsonDynamicValue)
```
If statically initialized it should be a FrozenDictionary.

sebastienros commented on src/OrchardCore/OrchardCore.Abstractions/Json/Dynamic/JsonDynamicValue.cs:
```diff
@@ -1,35 +1,297 @@
             {
-                _value = JsonValue?.GetObjectValue();
-                _hasValue = true;
+                var convertExpression = Expression.Convert(Expression.Convert(Expression, typeof(JsonDynamicValue)), targetType, castMethod);
```
Shouldn't this expression be cached instead since it varies only by type?

If so you can move the cache initialization code in a static constructor or it might involve unreadable code.

MikeAlhayek replied:
@gvkries are you able to address the last comment so we can merge this?

gvkries replied:
These expressions are wrapping another expression from the instance property `Expression`. We cannot cache them here.

sebastienros commented:
@MikeAlhayek saw that the cached methods were never used while debugging and fixed the issue. I get the point with the Expression. Also reformatted the methods with bodies because I could not read it while working on it locally (to the point that I was confusing the ctor with a property accessor).

sebastienros commented:
@gvkries what is the point of `JsonDynamicValue.Value`, it's only used in a test. I removed it but want to be sure it's not used "dynamically" for another reason.

gvkries commented:
> @gvkries what is the point of `JsonDynamicValue.Value`, it's only used in a test. I removed it but want to be sure it's not used "dynamically" for another reason.

I found a regression because formattable `ToString()` methods were used somewhere in a view. It's hard to find those references because of dynamic typing. Thats why I added `Value`back too, because it may also be used by someone in a template only. 
With Newtonsoft Json either `contentItem.Content.MyPart.myField.Text.Value` or `contentItem.Content.MyPart.myField.Text` was possible as well.

Therefore I think we should keep it.

MikeAlhayek commented:
> Also reformatted the methods with bodies because I could not read it while working on it locally

@sebastienros typically we use expression-bodies method when there is a single line of code. In this case, these are would be expression-bodied methods.

MikeAlhayek commented:
> If the goal is readability, then I swear it's doing the opposite in these cases. 

No strong opinion here beside that seems to be the rule we follow. Maybe we should suggest against the use of bodied expressions for methods? This way we can follow a simple rule across the board.

gvkries commented:
> > a single line of code
> 
> My feedback:
> 
> From this experience I believe that it shouldn't be the complete rule. I'd prefer something like "if the whole thing, method/property declaration "plus" implementation fits in a single line of code (< N char)."
> 
> If the goal is readability, then I swear it's doing the opposite in these cases. It's still going on a new line, but now you see a bunch of line where before you have braces to delimit the declaration from the body. And I had issues determining it was a method, or even in the case I mentioned a constructor.

I second that and if you don't mind, let me add one additional exception: Keep files consistent. E.g. when adding new methods either stick to the existing style or at least change the whole file (I would rather keep the existing format when making functional changes).

gvkries commented:
@sebastienros Just to make sure, you still don't want the `Value` property?

sebastienros commented:
> With Newtonsoft Json either contentItem.Content.MyPart.myField.Text.Value or contentItem.Content.MyPart.myField.Text was possible as well.

So the dynamic wrapper in Newtonsoft has a similar `Value` property? But we don't have tests for it, and we don't document it?

Honestly a text field has a `string Text` property and `contentItem.Content.MyPart.myField.Text` would just access it and work fine. If someone found this property (it's a documented one) I don't see why they would also type `.Value` after.

It might not hurt but once we start adding known/documented things it's hard to remove it so that's why I often try to push hard on not adding seemingly harmless things. I didn't understand the case that would not work without `.Value`, could you provide another example?

gvkries commented:
Yes, Newtonsoft has this `Value` property. I added that here just for backward compatibility, beside that I see no further need or justification for it. I only wanted to be sure it's not an oversight during merging.

sebastienros commented:
I checked the code base for .Value in a cshtml (where I assume it would be used) and only found the properties named `Value` in current fields, so I think it's ok to keep it this way. Thanks for the detailed investigation.
